core/test/mpi/base/bindings.cpp

```diff
@@ -0,0 +1,782 @@
+protected:
+    MpiBindings() : ref(gko::ReferenceExecutor::create()) {}
+
+    std::shared_ptr<gko::Executor> ref;
+
+    void assert_equal_arrays(gko::Array<double>& array_1,
+                             gko::Array<double>& array_2)
+    {
+        ASSERT_EQ(array_1.get_num_elems(), array_2.get_num_elems());
+        for (gko::size_type i = 0; i < array_1.get_num_elems(); ++i) {
+            EXPECT_EQ(array_1.get_const_data()[i], array_2.get_const_data()[i]);
+        }
+    }
+};
+
+
+TEST_F(MpiBindings, CanSetADefaultWindow)
+{
+    gko::mpi::window<int> win;
+    ASSERT_EQ(win.get(), MPI_WIN_NULL);
+}
+
+
+TEST_F(MpiBindings, CanCreateWindow)
+{
+    using ValueType = int;
+    ValueType* data;
+    data = new ValueType[4]{1, 2, 3, 4};
+    auto comm = gko::mpi::communicator::create_world();
+    auto win = gko::mpi::window<ValueType>(data, 4 * sizeof(ValueType), comm);
+    ASSERT_NE(win.get(), MPI_WIN_NULL);
+    win.lock_all();
+    win.unlock_all();
+    delete data;
+}
+
+
+TEST_F(MpiBindings, CanSendAndRecvValues)
+{
+    using ValueType = int;
+    auto comm = gko::mpi::communicator::create_world();
+    auto my_rank = comm->rank();
+    auto num_ranks = comm->size();
+    auto send_array = gko::Array<ValueType>{ref};
+    auto recv_array = gko::Array<ValueType>{ref};
+    ValueType* data;
+    if (my_rank == 0) {
+        data = new ValueType[4]{1, 2, 3, 4};
+        send_array =
+            gko::Array<ValueType>{ref, gko::Array<ValueType>(ref, 4, data)};
+        for (auto rank = 0; rank < num_ranks; ++rank) {
+            if (rank != my_rank) {
+                gko::mpi::send<ValueType>(send_array.get_const_data(), 4, rank,
+                                          40 + rank, comm);
+            }
+        }
+    } else {
+        recv_array = gko::Array<ValueType>{ref, 4};
+        gko::mpi::recv<ValueType>(recv_array.get_data(), 4, 0, 40 + my_rank,
+                                  comm);
+    }
+    if (my_rank != 0) {
+        ASSERT_EQ(recv_array.get_data()[0], 1);
+        ASSERT_EQ(recv_array.get_data()[1], 2);
+        ASSERT_EQ(recv_array.get_data()[2], 3);
+        ASSERT_EQ(recv_array.get_data()[3], 4);
+    }
+}
+
+
+TEST_F(MpiBindings, CanNonBlockingSendAndNonBlockingRecvValues)
+{
+    using ValueType = int;
+    auto comm = gko::mpi::communicator::create_world();
+    auto my_rank = comm->rank();
+    auto num_ranks = comm->size();
+    auto send_array = gko::Array<ValueType>{ref};
+    auto recv_array = gko::Array<ValueType>{ref};
+    ValueType* data;
+    auto req = gko::mpi::request::create(num_ranks);
+    if (my_rank == 0) {
+        data = new ValueType[4]{1, 2, 3, 4};
+        send_array =
+            gko::Array<ValueType>{ref, gko::Array<ValueType>(ref, 4, data)};
+        for (auto rank = 0; rank < num_ranks; ++rank) {
+            if (rank != my_rank) {
+                gko::mpi::send<ValueType>(send_array.get_data(), 4, rank,
+                                          40 + rank, req, comm);
```

Comment:
I believe this is a mistake? You're passing the same `MPI_Request` to all the sends. See my comments in the include file on that.